File: medium/113_path_sum_II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> res;
        if (!root) return res;
        vector<int> path;
        findPathWithSum(root, sum, res, path);
        return res;
    }
    
    void findPathWithSum(TreeNode* node, int sum, vector<vector<int>> &res, vector<int> &path) {
        path.push_back(node -> val);
        if (!node->left && !node->right && node->val == sum) {
            res.push_back(path);
        }
        // the key is to deduct node->val from sum!!
        if (node->left) 
            findPathWithSum(node->left, sum - node->val, res, path);
        if (node->right)
            findPathWithSum(node->right, sum - node->val, res, path);
        path.pop_back();
    }
};

File: medium/113_path_sum_II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    // DFS
    // Time: O(n)
    private List<List<Integer>> res;
    
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        res = new ArrayList<>();
        helper(root, 0, sum, new ArrayList<>());
        return res;
    }
    
    private void helper(TreeNode node, int curSum, int target, List<Integer> list) {
        if (node == null) return;
        int sum = curSum + node.val;
        list.add(node.val);
        if (node.left == null && node.right == null) {
            if (sum == target) {
                // Remember to use a deep clone!
                res.add(new ArrayList<>(list));  
            }
            list.remove(list.size()-1);
            return;
        }
        helper(node.left, sum, target, list);
        helper(node.right, sum, target, list);
        // Remember to reset!
        list.remove(list.size()-1);
    }
}